c10/core/impl/TorchDispatchModeTLS.cpp

```diff
@@ -10,40 +11,158 @@ namespace impl {
 
 thread_local TorchDispatchModeTLS torchDispatchModeState;
 
-void TorchDispatchModeTLS::push_onto_stack(std::shared_ptr<SafePyObject> mode) {
-  if (torchDispatchModeState.stack_.empty()) {
+bool TorchDispatchModeTLS::any_modes_set() {
+  if (!torchDispatchModeState.stack_.empty())
+    return true;
+  for (const auto i :
+       c10::irange(static_cast<size_t>(TorchDispatchModeKey::NUM_MODE_KEYS))) {
+    if (torchDispatchModeState.infra_modes_[i] != c10::nullopt) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void TorchDispatchModeTLS::push_onto_stack(
+    std::shared_ptr<SafePyObject> mode,
+    c10::optional<TorchDispatchModeKey> mode_key) {
+  if (!any_modes_set()) {
     c10::impl::tls_set_dispatch_key_included(DispatchKey::Python, true);
     c10::impl::tls_set_dispatch_key_included(
         DispatchKey::PythonTLSSnapshot, true);
   }
-  torchDispatchModeState.stack_.push_back(std::move(mode));
+  if (mode_key == c10::nullopt) {
+    torchDispatchModeState.stack_.push_back(std::move(mode));
+  } else {
+    set_mode(mode, mode_key.value());
+  }
 }
 
-const std::shared_ptr<SafePyObject> TorchDispatchModeTLS::pop_stack() {
-  TORCH_CHECK(
-      !torchDispatchModeState.stack_.empty(),
-      "trying to pop from empty mode stack");
-  std::shared_ptr<SafePyObject> out = torchDispatchModeState.stack_.back();
-  torchDispatchModeState.stack_.pop_back();
+const std::
+    tuple<std::shared_ptr<SafePyObject>, c10::optional<TorchDispatchModeKey>>
+    TorchDispatchModeTLS::pop_stack(
```

Comment:
Does this even still make sense.  It feels like you want something more similar to how regular dispatch keys work, where you say "I want to run the keys AFTER this one". Yolo popping a mode off the stack seems error prone, especially now that you have both a stack as well as infra modes

Reply:
There are currently only two usages of this API:

(1) `StashTorchDispatchModeGuard`, which needs to pop the highest mode temporarily off the stack during python dispatch, and then push that same mode back onto the stack on exit (making sure to not blindly push that mode to the **top** of the stack: if the mode was an infra mode, we need to set it with its corresponding mode key).

(2) The python-bound function used by `TorchDispatchMode.__exit__`, `pop_torch_dispatch_stack` (although in that function, I have a mode key so I know if I'm popping the top of the stack, vs. removing a specific infra mode).

I'm not sure that those two cases fit the "run the keys AFTER this one" pattern from dispatch keys. But I agree that this function on its own is a footgun. I can probably re-implement the above two usages without this function.

Right now, `push_onto_stack` and `pop_stack` know how to handle **both** user and infra modes - what do you think of the following instead?
```
# **only** pushes user modes
void pop_stack(std::shared_ptr<SafePyObject> mode)

# **only** pops user modes
std::shared_ptr<SafePyObject> pop_stack()

# **only** pops infra modes
# downside is that the two users above need to query the state and manually to see if there are any user modes first, that way they know whether or not to call `pop_stack` or `pop_highest_infra_mode`
std::tuple<std::shared_ptr<SafePyObject>, TorchDispatchModeKey> TorchDispatchModeTLS::pop_highest_infra_mode()
```

Reply:
I ended up doing the above - except, `pop_stack()` is allowed to first try to pop any user modes, and if there are none, also try to pop any infra modes (in priority order). This made it easier to keep the existing implementations of some of our python utils like `_disable_current_modes`.